maven-model-builder/src/main/java/org/apache/maven/model/building/AbstractModelSourceTransformer.java

```diff
@@ -0,0 +1,196 @@
+        }
+        catch ( TransformerConfigurationException | SAXException | ParserConfigurationException e )
+        {
+            throw new org.apache.maven.model.building.TransformerException( e );
+        }
+        
+        final SAXSource transformSource =
+            new SAXSource( filter,
+                           new org.xml.sax.InputSource( new FileInputStream( pomFile.toFile() ) ) );
+
+        final PipedOutputStream pout = new PipedOutputStream();
+        final PipedInputStream pipedInputStream = new PipedInputStream( pout );
+        
+        OutputStream out = filterOutputStream( pout, pomFile );
+
+        final javax.xml.transform.Result result; 
+        if ( transformerHandler == null )
+        {
+            result = new StreamResult( out );
+        }
+        else
+        {
+            result = new SAXResult( transformerHandler );
+            ( (SAXResult) result ).setLexicalHandler( new CommentRenormalizer( transformerHandler ) );
+            transformerHandler.setResult( new StreamResult( out ) );
+        }
+
+        IOExceptionHandler eh = new IOExceptionHandler();
+
+        Thread transformThread = new Thread( () -> 
+        {
+            try ( PipedOutputStream pos = pout )
+            {
+                transformerFactory.newTransformer().transform( transformSource, result );
+            }
+            catch ( TransformerException | IOException e )
+            {
+                eh.uncaughtException( Thread.currentThread(), e );
+            }
+        }, "TransformThread-" + TRANSFORM_THREAD_COUNTER.incrementAndGet() );
+        transformThread.setUncaughtExceptionHandler( eh );
+        transformThread.setDaemon( true );
+        transformThread.start();
+
+        return new ThreadAwareInputStream( pipedInputStream, eh );
+    }
+
+    private static class IOExceptionHandler
+        implements Thread.UncaughtExceptionHandler, AutoCloseable
+    {
+        private volatile Throwable cause;
+
+        @Override
+        public void uncaughtException( Thread t, Throwable e )
+        {
+            try
+            {
+                throw e;
+            }
+            catch ( TransformerException | IOException | RuntimeException | Error allGood )
+            {
+                // all good
+                this.cause = e;
+            }
+            catch ( Throwable notGood )
+            {
+                throw new AssertionError( "Unexpected Exception", e );
+            }
+        }
+
+        @Override
+        public void close()
+            throws IOException
+        {
+            if ( cause != null )
+            {
+                try
+                {
+                    throw cause;
+                }
+                catch ( IOException | RuntimeException | Error e )
+                {
+                    throw e;
+                }
+                catch ( Throwable t )
+                {
+                    throw new AssertionError( "Failed to transform pom", t );
```

Comment:
I don't think it is appropriate to use `Error` for use. I'd wrap into an runtime or illegal state exception.

Reply:
I think the `Error` makes sense. In case an OutOfMemoryError occurs during the transformation, you want to throw it as is, not wrap it.